File: Monitor_c/PitchFreq.c
#include "PitchFreq.h"
#include "max.h"
#include "smoothing.h"
#include "clipping.h"
#include "enframe.h"
#include <malloc.h>

void PitchFreq(double data[], int size, int fs, int nBit, int *feature_PitchFreq, double *feature_PitchFreq2)
{
	int i,j,k,m,n;

	double secPerFra = 30*0.001;
	int frameLength = secPerFra*fs;
	int frameInc = frameLength/3;
	int frameNum;
	int cutting;
	int Rmax_pos;
	int feature1=0;
	double *firData, *clippedData;
	double frameSTE, Rmax;
	double **frames;
	double *Rm;
	double *PitchFreq, *smoothedPitchFreq;
	int feature_PitchFreq2_count =0;
	*feature_PitchFreq2=0;

	Rm = (double *)malloc(sizeof(double)*frameLength);
	//firData = (double *)malloc(sizeof(double)*size);
	clippedData = (double *)malloc(sizeof(double)*size);

	//fir(data,firData);
	clipping(data,size,1,&clippedData);	
	
	enframe(clippedData, size, frameLength, frameInc, &frames, &frameNum);
	PitchFreq = (double *)malloc(sizeof(double)*frameNum);
	smoothedPitchFreq = (double *)malloc(sizeof(double)*frameNum);
	
	cutting = (int)(0.003*fs);	

	for(i=0;i<frameNum;i++)
	{
		frameSTE = 0;
		for(j=0;j<frameLength;j++)
			frameSTE += frames[i][j]*frames[i][j];	
		for(k=0;k<frameLength;k++)
		{
			Rm[k]=0;
			for(m=k+1;m<frameLength;m++)
				Rm[k] = Rm[k]+frames[i][m]*frames[i][m-k];
		}
		Rmax = max(Rm,frameLength,cutting,-1,&Rmax_pos);	
		if(Rmax<=0.25*frameSTE || Rmax==0)
			PitchFreq[i]=0;
		else
		{
			PitchFreq[i]=fs/(Rmax_pos+1);
		}
	}	

	smoothing(PitchFreq,frameNum,&smoothedPitchFreq);
	
	for(n=0;n<frameNum;n++)
	{
		if(smoothedPitchFreq[n]>200.00)
			feature1++;
		if(smoothedPitchFreq[n]>0)
			feature_PitchFreq2_count++;
	}
	*feature_PitchFreq = feature1;
	*feature_PitchFreq2 = (double)feature1/feature_PitchFreq2_count;

	free(Rm);
	free(clippedData);
	free(PitchFreq);
	free(smoothedPitchFreq);
}

File: matlab/PitchFreq.m
function [ feature_PitchFreq feature_PitchFreq2 ] = PitchFreq( voice, fs, nBit )
%PITCHFREQ Calc the pitch frequency feature of the voice
%   input: voice data
%   input: sample rate
%   input: number of bits per sample
%   output: pitch frequency feature of the voice data

    disp('Feature PitchFreq');
    voice = FIR(voice, fs, nBit, 3000);     %low pass filter (3000Hz)
    voice = clipping(voice, '3level');      %clipping (three level clipping 三电平削波)
    
    frameLength = 0.03*fs;  %根据采样频率选择帧长
    frameInc = frameLength/3;   %帧移为帧长1/3
    voiceFrames = enframe(voice, frameLength, frameInc);
    [framesNum frameLen] = size(voiceFrames); 
%     size(voiceFrames)
%     disp(framesNum);


    cutting = floor(0.003*fs);
    for m=1:framesNum;           %对每一帧求短时自相关函数
        Rm0=sum(voiceFrames(m,:).^2);   %STE of the frame
        for k=1:frameLength;
            Rm(k)=0;
            for i=(k+1):frameLength;
                Rm(k)=Rm(k)+voiceFrames(m,i)*voiceFrames(m,i-k);
            end
        end
        p=Rm(cutting:frameLength);                %防止误判，去掉前边3ms*fs个数值较大的点  
%         if sum(p~=0)~=0
%             figure(m);
%             plot(p);
%         end
        [Rmax,N(m)]=max(p);        %读取第一个自相关函数的最大点
        if Rmax<0.25*Rm0 || Rmax==0          %如果最大值小于STE的25%或者STE为0，则认为是无声的（或者是清音信号），基音周期置0
            N(m)=0;
        else
            N(m)=N(m)+cutting;
            N(m)=fs./N(m);
        end
    end 
%      N=N+cutting;
    T=N;
%      figure(1);stem(T,'.');axis([0 length(T) 0 max(T)]);
%      xlabel('帧数(n)');ylabel('周期(ms)');title('各帧基音周期');
    
    pitchFreq = smoothing(T);
%       figure(2);
     plot(pitchFreq);
    
    feature_PitchFreq = sum(pitchFreq>200);     %we're interested in frames whose pitch frequecy is bigger than 200
    feature_PitchFreq2 = sum(pitchFreq>200)/sum(pitchFreq~=0);
    
end

